src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -170,53 +169,45 @@ public void close() {
     ) {
         // if it's a starting clause
         if (this._child == null) {
-            // create initial RDD from expression
-            JavaRDD<Item> initialRdd = _expression.getRDD(context);
-
-            // define a schema
-            List<StructField> fields = new ArrayList<>();
-            StructField field = DataTypes.createStructField(_variableName, DataTypes.BinaryType, true);
-            fields.add(field);
-            StructType schema = DataTypes.createStructType(fields);
-
-            JavaRDD<Row> rowRDD = initialRdd.map(new ForClauseSerializeClosure());
-
-            // apply the schema to row RDD
-            return SparkSessionManager.getInstance().getOrCreateSession().createDataFrame(rowRDD, schema);
+            return getDataFrameFromRDDExpression(context);
         }
 
         if (_child.isDataFrame()) {
```

Comment:
The old code should still be kept, but with the additional test that the expression depends on a variable bound by the current FLWOR (and throw an error if _expression.isRDD())

The code on the right should also be kept, but with the additional test that the expression does not depend on a variable bound by the current FLWOR.

Testing whether the expression does or not depend on a variable bound by the current FLWOR should be straightforward, because there are a few methods in clauses and iterator classes that return the variable dependencies. It is just a test whether an intersection is empty or not.

Reply:
I will take a look at this now

Reply:
>(and throw an error if _expression.isRDD())

I remember you mentioning a custom Sparksoniq error message for this. Did you have a specific message in mind ?

Reply:
Yes, I suggest increasing (0003). However, I would suggest replacing SPRKIQ0001 and SPRKIQ0002 with RBST0001, RBST0002 calling the new one RBST0003 to stick with error code conventions from the W3C (2 letters for technology, 2 letters for static/dynamic/typed/etc and 4 digits). Does it help?